<a href="https://colab.research.google.com/github/Fellipe-A/Estudo_Python/blob/master/Python_zero2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<small><small><i>
Este notebook foi baseado nos notebooks disponíveis em [https://gitlab.erc.monash.edu.au/andrease/Python4Maths.git]
</i></small></small>

# Comparando e decidindo

Python permite comparar os valores associados a diferentes variáveis usando **operadores de comparação**. 

Expressões que contêm operadores de comparação são chamadas **expressões lógicas** e podem ser usadas em **estruturas de decisão** para decidir o que o seu código deve fazer. 

Neste roteiro, você aprenderá a utilizar estes conceitos da linguagem Python.

## Operadores de comparação

Operadores de comparação são utilizados para construir **expressões lógicas**. 

As regras básicas sobre expressões lógicas em Python são:

* Expressões lógicas são avaliadas como verdadeiras (**True**) ou falsas (**False**).
* Todos os operadores de comparação em Python têm precedência igual. 
* Assim como em expressões aritméticas, é possível usar parênteses para mudar a ordem de avaliação de uma expressão.
* Caso haja empate de precedência, a expressão é avaliada da esquerda para a direita.

Alguns dos operadores de comparação disponíveis em Python estão listados abaixo. 

| Símbolo | Significado |
|:----:|---|
| == | Igualdade |
| !=  | Diferença |
| < | Menor |
| > | Maior |
| <=  | Menor ou igual |
| >=  | Maior ou igual |

Note que é diferente usar `==` (comparação de igualdade) e `=` (associação de variável a valor).

In [0]:
z = 2
z == 2

True

In [0]:
z > 2

False

In [0]:
z <= 2

True

In [0]:
z != 2

False

### Exercícios de fixação

1 - Leia dois números inteiros e teste se o primeiro é menor do que o segundo.

In [0]:
n1 = int(input("Insira um valor: "))
n2 = int(input("Insira outro valor: "))

n1 < n2

Insira um valor: 3
Insira outro valor: 5


True

2 - Sorteie dois números e teste se ambos são iguais.

**Dica --** O procedimento **randint**(*menor*, *maior*) sorteia inteiros no intervalo entre menor e maior. Este procedimento não é um procedimento nativo do Python, por isso precisamos importá-lo do módulo **random**, que o fornece.

In [0]:
from random import randint
n1 = randint(0,100)
n2 = randint(0,100)
print(n1)
print(n2)
teste = (n1 == n2)
teste


54
7


False

### Comparando textos

Também é possível utilizar operadores de comparação para comparar textos:
* A igualdade (==) e a diferença (!=) funcionam de forma trivial, assim como no caso dos valores numéricos.
* As demais operações consideram a ordem alfabética.


In [0]:
nome = "Stevejobsson"
nome == "Stevejobsson"

True

In [0]:
nome != "Facebookson"

True

In [0]:
'abc' < 'bbc'

True

In [0]:
'bbc' <= 'bbc'

True

In [0]:
'bbc' <= 'bac'

False

### Exercícios de fixação

3 - Leia dois nomes e teste se o primeiro é anterior ao segundo considerando a ordem alfabética.

In [0]:
name1 = input("Nome 1: ")
name2 = input("Nome 2: ")

name1 <= name2

Nome 1: bbb
Nome 2: aaa


False

4 - Teste se os textos "Python" e "python" são considerados idênticos.

In [0]:
"Python" == "python"

False

## Operadores lógicos

É possível criar expressões lógicas mais complexas utilizando **operadores lógicos**.

Os operadores lógicos mais simples do Python são os operadores E (**and**) e OU (**or**).

As regras básicas sobre expressões que contenham operadores lógicos em Python são:

* Operadores de comparação têm maior precedência que operadores lógicos.
* O operador lógico **and** tem maior precedência que o operador **or**. 
* É possível usar parênteses para mudar a ordem de avaliação de uma expressão.
* Caso reste apenas operações de mesma precedência, a expressão passa a ser avaliada da esquerda para a direita.

| Símbolo | Significado |
|:----:|---|
| **and** | **True**, desde que ambos os lados da expressão sejam verdadeiros |
| **or**  | **True**, desde que pelo menos um lado da expressão seja verdadeiro |

In [0]:
"a" < "b" and 5 >= 5

True

In [0]:
"a" <= "b" and 5 > 5

False

In [0]:
"a" <= "b" or 5 > 5

True

### Exercícios de fixação

5 - Leia um gênero e uma idade. Teste se esta combinação representa alguém em situação de obrigatoriedade militar.

**Dica --** Em tempos de "paz", o serviço militar só é obrigatório para homens maiores de idade.

In [0]:
genero = input("Gênero: ")
idade = input("Idade: ")
if genero == "M" and int(idade) >= 18:
  print("Serviço militar obrigatório")
else:
  print("Serviço não obrigatório")

Gênero: F
Idade: 20
Serviço não obrigatório


6 - Leia uma idade e teste se esta pessoa está obrigada a votar.

**Dica --** Estão obrigados a votar pessoas entre 18 e 70 anos de idade.

In [0]:
idade = input("Idade: ")
if int(idade) >= 18 and int(idade) <= 70:
  print("Voto obrigatório")
else:
  print("Voto não obrigatório")

Idade: 15
Voto não obrigatório


7 - Leia uma idade e teste se para esta pessoa o voto é facultativo.

**Dica --** O voto é facultativo para pessoas entre 16 e 18 anos ou acima dos 70 anos.

In [0]:
idade = input("Idade: ")
if (int(idade) >= 16 or int(idade) < 18 ) or int(idade) >= 70:
  print("Voto facultativo")
else:
  print("Voto não facultativo")

Idade: 18
Voto facultativo



8 - Leia três inteiros e teste se o primeiro pertence ao intervalo delimitado pelos dois últimos.

**Observação --** Por simplicidade, suponha que os dois últimos números estão ordenados de forma crescente.

In [0]:
n = [5, 1, 10]

if n[0] >= n[1] and n[0] <= n[2]:
    print("O numero pertence ao intervalo")
else:
  print("O numero nao pertence ao intervalo")

O numero pertence ao intervalo


5

### Estruturas de decisão

A principal estrutura utilizada em Python para definir se um trecho de um código será executado é a estrutura **if**, ilustrada a seguir:

```python
if condicao:
    expressao
    expressao
    ...```

As regras básicas para o uso da estrutura **if** em Python são:

* O trecho de código delimitado por uma estrutura **if** é chamado bloco.
* Um bloco é caracterizado pela *identação* de suas linhas. Assim, a quantidade de espaços no começo de cada linha é bastante importante no Python.
* O bloco de uma estrutura **if** só é executado se sua condição for satisfeita. A condição é avaliada como uma expressão lógica.

In [0]:
numero = int(input())
if numero % 2 == 0:
    print("O número", numero, "é par.")
else:
  print("O numero", numero, "é impar.")

17
O numero 17 é impar.


### Exercícios de fixação

9 - Leia um inteiro e o imprima se ele for múltiplo de 3.

In [0]:
numero = int(input("Insira um valor: "))
if numero %3 == 0:
  print("O numero %d é múltiplo de 3" % numero)
else: 
  print("O numero %d não é multiplo de 3" % numero)

Insira um valor: 20
O numero 20 não é multiplo de 3


10 - Informe entradas para o código abaixo que façam a mensagem ser impressa.

In [0]:
texto = input()
numero = int(input())
if texto < "hello" and numero ** 0.5 == 8:
    print("Mensagem impressa ;)")

aa
64
Mensagem impressa ;)


### A estrutura ```if-else```

Uma expansão da estrutura **if** é a estrutura **if-else**, que define trechos de código mutuamente exclusivos.

```python
if condicao:
    # bloco se a condição for satisfeita
    expressao
    ...
else:
    # bloco se a condição não for satisfeita
    expressao
    ...
```
Ambos os blocos de uma estrutura **if-else** devem estar *identados*, de preferência pela mesma quantidade de espaços.

In [0]:
numero = int(input())
if numero % 2 == 0:
    print("O número", numero, "é par.")
else:
    print("O número", numero, "é ímpar.")

2578
O número 2578 é par.


### Exercícios de fixação

11 - Leia um inteiro e informe se ele é ou não múltiplo de 6.

In [0]:
numero = int(input("Insira um valor: "))
if numero %6 == 0:
  print("O numero %d é múltiplo de 6" % numero)
else: 
  print("O numero %d não é multiplo de 6" % numero)

12 - Leia dois nomes de pessoas e os imprima em ordem alfabética.

In [0]:
nome1 = input("Informe um nome: ")
nome2 = input("Informe outro nome: ")
if nome1 < nome2:
  print("%s %s" %(nome1, nome2))
else:
  print("%s %s" %(nome2, nome1))

Informe um nome: vv
Informe outro nome: as
as vv


### A estrutura ```if-elif-else```

É possível utilizar a estrutura **if-elif-else** para especificar múltiplos fluxos de execução do algoritmo, em função de uma condição principal e de condições alternativas:

```python
if condicao_principal:
    # bloco se a condição principal for satisfeita
    expressao
    ...
elif condicao_alternativa_1:
    # bloco se a condição principal não for satisfeita mas a condição alternativa 1 for
    expressao
    ...
elif condicao_alternativa_2:
    # bloco se nem a condição principal nem a condição alternativa 1 forem satisfeitas, mas a condição alternativa 2 for
    expressao
    ...
...
else:
    # bloco se nem a condição principal nem as condições alternativas forem satisfeitas
    expressao
    ...

```

As regras básicas para o uso da estrutura **if-elif-else** em Python são:

* O bloco **if** é executado se a condição principal for satisfeita.
* Um bloco **elif** é executado se sua condição for satisfeita e as condições acima dele não forem. 
* O bloco **else** é executado se nenhuma condição for satisfeita.


In [0]:
from random import randint

numero = randint(0,100)
chute = int(input("Chute um número entre 0 e 100: "))
if chute == numero:
    print("Parabéns, você acertou!")
elif chute < numero:
    print("Seu chute foi menor que o número sorteado!")
else:
    print("Seu chute foi maior que o número sorteado!")

Chute um número entre 0 e 100: 15
Seu chute foi menor que o número sorteado!


### Exercícios de fixação

13 - Leia um número e teste se ele é positivo, negativo ou nulo.

In [0]:
chute = int(input("Informe um valor: "))
if chute == 0:
    print("Número nulo!")
elif chute < 0:
    print("Número negativo!")
else:
    print("Número positivo!")

Informe um valor: -54
Número negativo!


### Combinando múltiplas estruturas de decisão

Também é possível combinar estruturas de decisão em múltiplos níveis, o que é conhecido como **aninhar estruturas**.

As regras básicas para aninhar estruturas de decisão em Python são:
* Não há limite com relação a quantas estruturas e em quantos níveis.
* Blocos de níveis mais profundos devem apresentar maior grau de identação. 

In [0]:
from random import randint

numero = randint(0,1000)
if numero % 6 == 0:
    print("O número", numero, "é múltiplo de 6!")
else:
    if numero % 2 == 0:
        print("O número", numero, "é múltiplo de 2, mas não de 3!")
    elif numero % 3 == 0:
        print("O número", numero, "é múltiplo de 3, mas não de 2!")
    else:
        print("O número", numero, "não é multiplo de 2 nem de 3, muito menos de 6!")

O número 654 é múltiplo de 6!


## Exercícios do URI

[1037](https://www.urionlinejudge.com.br/judge/pt/problems/view/1037) - Um histograma é um tipo de gráfico estatístico que particiona os dados disponíveis em subgrupos e informa quantos elementos existem em cada subgrupo.

Escreva um código Python que possa ser usado para construir um histograma de um conjunto de números inteiros sorteados no intervalo [0,100), isto é, valores maiores ou iguais a 0 e menores que 100. 

Por simplicidade, considere que apenas um número será sorteado. Você deve identificar a qual dos intervalos a seguir este número pertence: [0,25), [25,50), [50,75), [75,100). 

|.| Entrada | Saída |
|-|:----:|---|
| *Exemplo 1* | 5 | [0,25) | 
| *Exemplo 2* | 50 | [50,75) |
| *Exemplo 3* | 74 | [50,75) |

In [0]:
from random import randint

numero = randint(0,100)
if numero >= 0 and numero < 25:
    print("[0,25)")
elif numero >= 25 and numero < 50:
    print("[25,50)")
elif numero >= 50 and numero < 75:
    print("[50,75)")
else:
    print("[75,100)")

[0,25)


[1049](https://www.urionlinejudge.com.br/judge/pt/problems/view/1049) - Uma árvore de classificação é uma ferramenta usada em diferentes áreas da ciência para ajudar a identificar objetos por suas características.

Considerando a árvore de classificação abaixo, escreva um código Python que identifique um animal por suas características. Mais precisamente, seu código deve receber três textos como entrada e informar ao usuário de qual animal se trata.

<img src="https://github.com/leobezerra/python-zero/blob/master/imgs/UOJ_1049_b.png?raw=1">

|.| Entrada | Saída |
|-|----|---|
| *Exemplo 1* | vertebrado<br/>mamifero<br/>onivoro | homem | 
| *Exemplo 2* | vertebrado<br/>ave<br/>carnivoro | aguia |
| *Exemplo 3* | invertebrado<br/>anelideo<br/>onivoro | minhoca |

[1041](https://www.urionlinejudge.com.br/judge/pt/problems/view/1041) - Em sistemas de computação gráfica, compreender diferentes sistemas de coordenadas é fundamental.

Considere o sistema de coordenadas cartesiano apresentado na figura abaixo. Ele serve para representar pontos que apresentam duas dimensões, normalmente representados com a notação (x, y), onde x representa a coordenada horizontal e y a vertical.

<img src="https://github.com/leobezerra/python-zero/blob/master/imgs/UOJ_1041.png?raw=1">

O ponto (0,0) representa a origem dos eixos e é a intersecção entre as quatro subregiões ilustradas na figura abaixo (Q1, Q2, Q3 e Q4, onde Q significa *quadrante*, isto é, uma subregião). Cada quadrante da figura representa pontos com as seguintes características:

| Quadrante | Horizontal | Vertical |
|:-:|:-:|:-:|
| Q1 | x > 0 | y > 0 |
| Q2 | x < 0 | y > 0 |
| Q3 | x < 0 | y < 0 |
| Q4 | x > 0 | y < 0 |

Também é possível que um ponto esteja sobre um dos eixos:
* Um ponto estará sobre o eixo x se y = 0;
* Um ponto estará sobre o eixo y se x = 0.

Considerando este contexto, escreva um código Python que leia um ponto (duas coordenadas reais) e informe a primeira condição abaixo que se mostrar verdadeira:
* Se o ponto está na "Origem";
* Se o ponto está sobre o "Eixo X";
* Se o ponto está sobre o "Eixo Y";
* Em qual quadrante o ponto está.

|.| Entrada | Saída |
|-|:-:|:-:|
| *Exemplo 1* | 4.5<br/>-2.2 | Q4 | 
| *Exemplo 2* | 0.1<br/>0.0 | Eixo X |
| *Exemplo 3* | 0.0<br/>0.0 | Origem |

[1051](https://www.urionlinejudge.com.br/judge/pt/problems/view/1051) - Sistemas contábeis e de recursos humanos precisam conhecer bem a legislação tributária do país para calcular os impostos a serem recolhidos mensalmente sobre o salário dos funcionários. 

Trabalhadores também precisam entender quanto ganham e porque existem descontos em seus contra-cheques.

Neste exercício, você vai ajudar empresas e trabalhadores a entender os descontos da previdência social (INSS).

A tabela usada para calcular esses descontos é apresentada a seguir.

| Salário de contribuição | Alíquota |
|:-:|:-:|
| Até 1.693,72 | 8% | 
| De 1.693,73 a 2.822,90 | 9% |
| De 2.822,91 até 5.645,80 | 11% |

Como você pode ver, o desconto sobre o salário depende da faixa salarial na qual um trabalhador se enquadre.

Além disso, existe um teto na tabela que determina que pessoas que ganhem a partir de 5.645,81 devem contribuir com 621,04 (11% de 5.645,80).

Escreva um código Python que receba um salário bruto (antes do desconto) e informe o valor da contribuição para o INSS e o salário líquido (após o desconto).

|.| Entrada | Saída |
|-|:-:|-|
| *Exemplo 1* | 7456.57 | 621.04 6835.53 | 
| *Exemplo 2* | 2822.90 | 254.06 2568.84 |
| *Exemplo 3* | 1693.73 | 152.44 1541.29 |

Um segundo tipo de dedução mensal que incide sobre os ganhos dos trabalhadores é o *imposto de renda sobre pessoa física* (IRPF).

Para o cálculo do IRPF, utiliza-se como base de cálculo o salário bruto descontado da contribuição do INSS. 

As principais diferenças em relação ao cálculo do INSS são:
* Quem possui salário-base até 1.903,98 é isento do pagamento do IRPF.
* Não há teto para a contribuição do IRPF.
* O valor do IRPF cobrado a uma pessoa é abatido de um valor que varia em função da alíquota correspondente.

A tabela usada para o cálculo do IRPF é mostrada abaixo.

| Base de cálculo (R\$) | Alíquota | Parcela a deduzir |
|:-:|:-:|:-:|
| Até 1.903,98 | - | - |
| De 1.903,99 até 2.826,65 | 7,5% | 142,80 |
| De 2.826,66 até 3.751,05 | 15% | 354,80 |
| De 3.751,06 até 4.664,68 | 22,5% | 636,13 |
| Acima de 4.664,68 | 27,5% | 869,36 |

Para entender melhor o cálculo do IRPF, vamos tomar como exemplo uma pessoa que, descontada a contribuição do INSS, ganha 2.000,00.

A alíquota para esta remuneração é de 7,5% e o abatimento de 142,80. Assim, a contribuição do IRPF será de 2.000,00 * 7,5% - 142,80 = 7,20.

Escreva um código Python que receba um salário-base (já descontado do INSS) e informe o valor da contribuição do IRPF e o salário líquido (após o desconto do IRPF).

|.| Entrada | Saída |
|-|:-:|-|
| *Exemplo 1* | 6835.53 | 1010.41 5825.12 | 
| *Exemplo 2* | 2568.84 | 49.86 2518.98 |
| *Exemplo 3* | 1541.29 | 0.00 1541.29 |